add env var support into executables

pkit opened the pull request:
### UPDATED:

- adds a new setting `ENVIRONMENT VAR1 = 'a', VAR2 = 'b'` to table
functions
- improves "set query" handler to work on any settings type, even partial
- adds alow list for env vars to config (with globbing support)

Multiline config:
```xml
    <user_scripts>
        <allow_env_variables>
            <VAR_*/>
            <VAR1/>
        </allow_env_variables>
    </user_scripts>
```
Single pattern config:
```xml
    <user_scripts>
        <allow_env_variables>*</allow_env_variables>
    </user_scripts>
```

Query syntax:

```sql
SELECT * FROM
  executable('<script name>', '<format>', '<columns>',
    (SELECT * FROM table),
    ENVIRONMENT VAR1 = 'test', VAR2 = 1,
    SETTINGS <.....>
  )
```

Note: max additional environment size is limited to 4KB

### DEPRECATED:

Syntax:
```sql
SELECT * FROM
  executable('<script name>', '<format>', '<columns>',
    (SELECT * FROM table),
    SETTINGS env.VAR1 = 'test', env.VAR2 = 1
  )
```
_note: will **not** overwrite existing env vars_

fixes: #39840

### Changelog category (leave one):
- New Feature


### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
environment variable support for executable table functions

kitaisreal commented:
@pkit you can pass parameters into function, should we add also env var support ?

pkit commented:
@kitaisreal I'm not sure how to pass params into function so it doesn't require escaping too much
Do you have an idea for a good syntax ?

kitaisreal commented:
@pkit we already have such functionality, it works only for `executable` functions with `execute_direct`. Check example in docs https://clickhouse.com/docs/en/sql-reference/functions/#executable-user-defined-functions.
Script will look like this:
```python3
#!/usr/bin/python3

import sys

if __name__ == "__main__":
    for line in sys.stdin:
        print("Parameter " + str(sys.argv[1]) + " value " + str(line), end="")
        sys.stdout.flush()
```
Query:
```sql
SELECT test_function_parameter_python(1)(2);
```
Result:
```
┌─test_function_parameter_python(1)(2)─┐
│ Parameter 1 value 2                  │
└──────────────────────────────────────┘
```

pkit commented:
@kitaisreal in each script we need to map position to key and it works essentially in UDF only, not in UDTF?
Dunno if it's good enough.
Env vars is a standard way to accept arguments, obviously we can generate wrapper scripts with argument setup that we need, but that's actually another exec call with all the costs...
Just consider the situation in the wild:
AWS Lambda - env var support
Cloudflare Workers - env var support

alexey-milovidov commented:
It's strange to put them inside the SETTINGS clause and introduce a special case for the syntax of their names.
Maybe make it with

`SETTINGS a = b ENVIRONMENT c = d`

?

alexey-milovidov commented:
Also, I suspect, this is dangerous.

We ensure that the user can only use executables specified in the configuration file.
But with env override, like LD_PRELOAD, you can do arbitrary code execution.

alexey-milovidov commented:
How to avoid this:
- add allow-list of environment variables (in the UDF config) that can be changed;
- do nothing, let all the variables passed uniformly via command line arguments, and let everything else be handled by a wrapper executable, configured by a user.

pkit commented:
> SETTINGS a = b ENVIRONMENT c = d

Yeah, I've also considered that one.
If it's better, I can do it.

> do nothing, let all the variables passed uniformly via command line arguments, and let everything else be handled by a wrapper executable, configured by a user.

Yup, I had this one in the first message. It means we are setting up a new process twice per each call to executable.
It's ok for a long running stuff, but for something pumped by blocks in real-time through materialized view... it looks suboptimal.
Will think about it.

What about running real tightened process when?
I.e. do something similar to what k8s or docker does for processes: clone mount points, pids, uids, uts etc. but not networking.
It should be probably another engine though.

The ultimate goal here is an ability to pump blocks in real time through some processing that does non-trivial transformation logic (DL/ML model, external service access, etc.) quite a lot of features are missing for that one and I will try to add them gradually.

alexey-milovidov commented:
> but not networking.
It should be probably another engine though.

Networking is needed for some UDFs: for example, DNS requests.

Let's add configurable allowed env variables per UDF.

pkit commented:
> Networking is needed for some UDFs

Here I mean that the namespace for networking will not change, as there is no point to mess with it.
Will be equivalent of `docker .... --network=host`
Essentially it's like running docker but with less restrictions, as a lot of them do not make sense for an executable that's fed from fd streams.

> Let's add configurable allowed env variables per UDF.

Can do it. Is support for globbing ok there? (i.e. `allowed_env_vars=*` or somthing)

Are you in favor of `ENVIRONMENT a=b` syntax?

alexey-milovidov commented:
> Can do it. Is support for globbing ok there?

Yes.

> Are you in favor of ENVIRONMENT a=b syntax?

Yes.

pkit commented:
@alexey-milovidov @kitaisreal I'm not sure where in config should I put the `allowed_env_variables` expression?
For UDF and UDict it's anyway configured from config only, but there is no specific config per UDTF?
We can allow env vars in config per UDF/UDict directly, I suppose?

alexey-milovidov commented:
Yes, inside the config of every function, so it can be tuned for every function separately.

pkit commented:
@alexey-milovidov I still cannot get the exact syntax, for example if the table function looks like:
```sql
SELECT * FROM
  executable('my_model.py predict', 'ArrowStream', 'prediction Float32',
    (SELECT * FROM table),
    ENVIRONMENT VAR1 = 'test', VAR2 = 1
  )
```
How exactly the entry in config file should look like?

alexey-milovidov commented:
Ok, I understand - executable table functions don't have per-function config entry.
Let's do something like:

```
<user_scripts>
    <allow_env_variables>
        <var1/>
        <var2/>
    </allow_env_variables>
</user_scripts>
```

- globally allowed env variables for table functions and UDFs.

pkit commented:
@kitaisreal @alexey-milovidov I'm hitting a wall here
I've tried to create a good parser for _partial settings_ by inheriting from the current "SetQuery" parser
```C++
class ASTSettings : public ASTSetQuery
{
public:
    explicit ASTSettings(const char * token_)
    {
        token = String(token_);
    }

    String getID(char delimiter) const override { return String("Set") + delimiter + token; }

    void formatImpl(const FormatSettings & format, FormatState &, FormatStateStacked) const override;

    String getToken() { return token; }
protected:
    String token;
};
```
And it works well in tests, everything is fine.
But when I get to actual `void TableFunctionExecutable::parseArguments(const ASTPtr & ast_function, ContextPtr context)` code the `args[i]->as<ASTSettings>()` is false.
**Does it mean that somewhere AST tree is reassembled again? And then the subclass information is lost...**

Simple test like that works fine (prints true):
```c++
int main(int, char **)
{
    using namespace DB;

    std::string input = "SELECT * FROM executable('arrow_sum.py', 'ArrowStream', 'sum UInt64', (SELECT * FROM numbers(10*1000000)), ENVIRONMENT VAR1 = 'a', SETTINGS command_termination_timeout = 100)";
    ParserQueryWithOutput parser(input.data() + input.size());
    ASTPtr ast = parseQuery(parser, input.data(), input.data() + input.size(), "", 0, 0);
    auto child = ast
                     ->children.at(0)
                     ->children.at(0)
                     ->children.at(1)
                     ->children.at(0)
                     ->children.at(0)
                     ->children.at(0);
     std::cout << !!child->as<ASTFunction>()->arguments->children.at(4)->as<ASTSettings>() << std::endl;
     return 0;
}
```

pkit commented:
Ah, found it! `ASTPtr clone() const override { return std::make_shared<ASTSetQuery>(*this); }` clone always called the parent...
Indeed the tree gets reassembled through cloning...

Commit:
implement env var support for Executable TF

- adds a new setting `ENVIRONMENT VAR1 = 'a', VAR2 = 'b'` to table
functions
- improves "set query" handler to work on any settings type, even partial
- adds alow list for env vars to config (with globbing support)

Multiline config:
```xml
    <user_scripts>
        <allow_env_variables>
            <VAR_*/>
            <VAR1/>
        </allow_env_variables>
    </user_scripts>
```
Single pattern config:
```xml
    <user_scripts>
        <allow_env_variables>*</allow_env_variables>
    </user_scripts>
```

pkit commented:
OK, looks like it's working as expected finally, all features discussed here implemented
Will add tests soon

pkit commented:
@kitaisreal @alexey-milovidov Done. Tests added, should be ready for review.

alexey-milovidov commented:
It looks like this PR was forgotten by @kitaisreal. Now it is ready for review.

alexey-milovidov commented:
This is a good PR, but it looks abandoned, and it has conflicts.

alexey-milovidov commented:
Please reopen when it is ready.

alexey-milovidov commented on src/Common/ShellCommand.cpp:
```diff
@@ -115,6 +116,17 @@ void ShellCommand::logCommand(const char * filename, char * const argv[])
         /// NOTE: No escaping is performed.
         args << "'" << argv[i] << "'";
     }
+    if (!env->empty())
```
`env` was passed by pointer, but wasn't checked for nullptr.

alexey-milovidov commented on src/Common/ShellCommand.cpp:
```diff
@@ -104,7 +105,7 @@ bool ShellCommand::tryWaitProcessWithTimeout(size_t timeout_in_seconds)
 }
 
-void ShellCommand::logCommand(const char * filename, char * const argv[])
+void ShellCommand::logCommand(const char * filename, char * const argv[], const std::vector<std::string> *env)
```
There is some inconsistency in the arguments of this function.

alexey-milovidov commented on src/Common/ShellCommand.cpp:
```diff
@@ -115,6 +116,17 @@ void ShellCommand::logCommand(const char * filename, char * const argv[])
+        {
+            if (i > 0)
+                args << ", ";
+            args << (*env)[i] << "='" << (*env)[i + 1] << "'";
```
The value isn't quoted and escaped.

alexey-milovidov commented on src/Common/ShellCommand.cpp:
```diff
@@ -115,6 +116,17 @@ void ShellCommand::logCommand(const char * filename, char * const argv[])
+    if (!env->empty())
+    {
+        args << " (env: ";
+        for (size_t i = 0; i < env->size(); i += 2)
```
The size is evaluated every time in a loop.

alexey-milovidov commented on src/Common/ShellCommand.cpp:
```diff
@@ -204,12 +218,17 @@ std::unique_ptr<ShellCommand> ShellCommand::executeImpl(
         _exit(static_cast<int>(ReturnCodes::CANNOT_EXEC));
     }
 
+    delete env;
```
Oops... after reading this, I have to stop and breathe.

alexey-milovidov commented on src/Common/ShellCommand.cpp:
```diff
@@ -335,5 +354,50 @@ void ShellCommand::wait()
+        orig_number_of_env_vars = env_num;
+        size_t needed_size = (env_num + env_vars->size() / 2 + 1) * sizeof(*envp);
+        data = static_cast<char **>(malloc(needed_size));
+        // here we only copy the pointers, as clickhouse should __never__ call setenv() when threads are up
```
`__never__` sounds scary.

alexey-milovidov replied:
`f()` is K&R style, a function with name `f` should be written as `f` in comments.
